core/templates/dev/head/pages/exploration-editor-page/improvements-tab/services/improvement-modal.service.ts

```diff
@@ -48,6 +48,181 @@ angular.module('oppia').factory('ImprovementModalService', [
+            'LearnerActionRenderService',
+            'ISSUE_TYPE_MULTIPLE_INCORRECT_SUBMISSIONS',
+            'MAX_UNRELATED_ACTIONS_PER_BLOCK',
+            function(
+                $scope, $uibModalInstance, AlertsService,
+                LearnerActionRenderService,
+                ISSUE_TYPE_MULTIPLE_INCORRECT_SUBMISSIONS,
+                MAX_UNRELATED_ACTIONS_PER_BLOCK) {
+              $scope.playthroughIndex = index;
+
+              // This is the index of the latest learner action that was
+              // displayed. When this index reaches 0, we've displayed all the
+              // learner actions.
+              let currActionIndex = playthrough.actions.length - 1;
+              // We keep track of the indices where we expanded till at each
+              // level so that we would be able to display the final block as
+              // a table in Multiple Incorrect Submissions issue.
+              let indicesToExpand = [];
+              let actionsToDisplay = [];
+
+              /**
+               * Identifies whether all learner actions have been displayed and
+               * no further expansion is possible.
+               * @returns {boolean}
+               */
+              $scope.canExpandActions = function() {
+                return currActionIndex !== 0;
+              };
+
+              /**
+               * Iterates through the learner actions and expands the set of
+               * actions to be displayed. If all learner actions have been
+               * handled, no more expansion is possible.
+               */
+              let expandActions = function() {
+                let i;
+                for (i = currActionIndex; i >= 0; i--) {
+                  let action = playthrough.actions[i];
+                  let currentStateName =
+                      action.actionCustomizationArgs.state_name.value;
+                  let latestStateName = actionsToDisplay.length === 0 ? (
+                      action.actionCustomizationArgs.state_name.value) : (
+                      actionsToDisplay[
+                        actionsToDisplay.length -
+                          1].actionCustomizationArgs.state_name.value);
+
+                  if (currentStateName !== latestStateName) {
+                    // When the latest state name and the state name of the
+                    // current learner action differ, they are part of
+                    // different contexts. So, we identify whether the size
+                    // of the block has crossed the threshold number of actions
+                    // per block. If it has, we don't add more learner actions
+                    // to the block.
+                    if (currActionIndex - i >= (
+                      MAX_UNRELATED_ACTIONS_PER_BLOCK)) {
+                      currActionIndex = i + 1;
+                      break;
+                    }
+                    actionsToDisplay.push(action);
```

Comment:
Don't need this if we're just referencing indexes. Ditto below.

Reply:
Done.